### PR TITLE
docs(reference): major docs refactor

### PR DESCRIPTION
**Pull request by steebchen**

### CODE REVIEW

File: docs/reference/01-introduction.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Introduction
+
+The Prisma Go client aims to be fully typesafe wherever possible, even for complex queries.
+That's why it uses a functional syntax to make sure every argument you provide matches the type in the database.
+If you change something in the database, and the change is incompatible with the code, you will get a compile time error.
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:5]: Great intro :)

File: docs/reference/01-introduction.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    content   String
+
+    post   Post @relation(fields: [postID], references: [id])
+    postID String
+}
+```
+
+## The syntax concept
+
+The generated go client uses a specific and consistent functional API syntax.
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:33]: go => Go

File: docs/reference/01-introduction.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+A simple line to fetch all posts could look as follows:
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:35]: For writing technical content, the css-tricks.com creator mentioned his writing "tone" on a podcast that stuck with me:

> Write like we're going on an adventure together.

Applying this to the above...

"A simple line to fetch all posts could look as follows:"

to

"To fetch all posts, you can do:"

File: docs/reference/01-introduction.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+```go
+posts, err := client.Post.FindMany().Exec(ctx)
+```
+
+Usually you want to query for some ID or other field. The go client exposes a completely typesafe query builder
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:41]: go => Go. typesafe => type-safe

File: docs/reference/01-introduction.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    // package.
+    // model.
+    //      field.
+    //            method.
+    //                  (value)
+    db.Post.Title.Equals("hi"),
+).Exec(ctx)
+```
+
+You can swap them out as you wish, given that the query is possible for whatever you're querying. For example, you can
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:55]: what's "them" refer to?

File: docs/reference/01-introduction.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+only query for fields which actually exist on a given model. If a field is optional, you will also get additional
+methods such as IsNull() and *Optional variations to query for sql NULLs:
```
Comment (matthewmueller) [inline docs/reference/01-introduction.md:57]: sql => SQL

File: docs/reference/02-find.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+### Find many records
+
+```go
+posts, err := client.Post.FindMany(
+    db.Post.Title.Equals("hi"),
+).Exec(ctx)
+```
+
+If no records are found, this returns an empty array without returning an error (like usual SQL queries).
```
Comment (matthewmueller) [inline docs/reference/02-find.md:39]: "this" is a bit ambiguous. Maybe "FindMany" or "this query" or "the above".

like usual SQL queries => like SELECT queries in SQL.

File: docs/reference/02-find.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+post, err := client.Post.FindOne(
+    db.Post.ID.Equals("123"),
+).Exec(ctx)
+
+if err == db.ErrNotFound {
+    log.Printf("no record with id 123")
+}
+```
+
+This returns an error of type `ErrNotFound` (exported in the `db` package) if there was no such record.
```
Comment (matthewmueller) [inline docs/reference/02-find.md:53]: returns an `ErrNotFound` type (exported **by** the `db` package).

File: docs/reference/02-find.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+Depending on the data types of your fields, you will automatically be able to query for respective operations. For example, for integer or float fields you might want to query for a field which is less than or greater than some number.
+
+```go
+post, err := client.Post.FindOne(
+    // query for posts containing the title "hi"
+    db.Post.Title.Contains("what up"),
+).Exec(ctx)
+```
+
+To explore more query filter, see [all possible query filters](./03-filters.md).
```
Comment (matthewmueller) [inline docs/reference/02-find.md:66]: filter => filters

File: docs/reference/02-find.md [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+if err == db.ErrNotFound {
+    log.Printf("no record with id 123")
+}
+```
+
+This returns an error of type `ErrNotFound` (exported in the `db` package) if there was no such record.
+
+### Query API
+
+Depending on the data types of your fields, you will automatically be able to query for respective operations. For example, for integer or float fields you might want to query for a field which is less than or greater than some number.
```
Comment (matthewmueller) [inline docs/reference/02-find.md:57]: A bit wordy, but I like the example. Maybe..

The query operations change based on the data types in your schema. For example, integers and floats will have greater than and less than operations, while strings have prefix and suffix operations.

File: docs/reference/03-filters.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
+# Query filters
+
+## Type filters
+
+When querying for data, you often need to query for specific things, such as if a database column contains a word,
```
Comment (matthewmueller) [inline docs/reference/03-filters.md:5]: "When querying for data, you often need to query for specific things"

This sentence offers no information... 😅 Maybe remove or rework.

File: docs/reference/03-filters.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+db.Post.CreatedAt.Before(time.Now().Truncate(24 * time.Hour)),
+// query for all posts which were created until yesterday including right now
+db.Post.CreatedAt.BeforeEquals(time.Now().Truncate(24 * time.Hour)),
+```
+
+## General
+
+### NOT
+
+If you want to negate a query, you can use NOT.
```
Comment (matthewmueller) [inline docs/reference/03-filters.md:59]: Not

File: docs/reference/03-filters.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+```go
+db.Post.Not(
+  db.Post.Title.Equals("123"),
+)
+```
+
+### Or
+
+If you want to negate a query, you can use NOT.
```
Comment (matthewmueller) [inline docs/reference/03-filters.md:71]: Or

File: docs/reference/05-create.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    // optional fields
+    Post.ID.Set("id"),
+    Post.Title.Set("name"),
+    Post.Stuff.Set("stuff"),
+).Exec(ctx)
+```
+
+### Create a record with a relation
+
+Use the method `Link` to connect new objects with existing ones. For example, the following query creates a new post and sets the postID attribute of the comment.
```
Comment (matthewmueller) [inline docs/reference/05-create.md:46]: Use the `Link` method. 

We've been using the word records to talk about results, so objects => records.

File: docs/reference/06-update.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
+# Create records
```
Comment (matthewmueller) [inline docs/reference/06-update.md:1]: Update records

File: docs/reference/05-create.md [hunk 16]
```diff
@@ -0,0 +1,83 @@
+# Create records
+
+Find, update and delete records.
```
Comment (matthewmueller) [inline docs/reference/05-create.md:3]: Find, create

File: docs/reference/07-delete.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
+# Create records
+
+Find, update and delete records.
```
Comment (matthewmueller) [inline docs/reference/07-delete.md:3]: Fix this line? I think this guide only talks about deleting. Same with the above.

File: docs/reference/07-delete.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
+# Create records
```
Comment (matthewmueller) [inline docs/reference/07-delete.md:1]: Delete

File: docs/reference/08-relations.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
+# Create records
```
Comment (matthewmueller) [inline docs/reference/08-relations.md:1]: Fix

File: docs/reference/08-relations.md [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+Find, update and delete records.
```
Comment (matthewmueller) [inline docs/reference/08-relations.md:3]: Fix

File: docs/reference/09-raw.md [hunk 21]
```diff
@@ -45,3 +45,7 @@ err := client.QueryRaw(`SELECT * FROM "User" WHERE id = $1 AND email = $2`, "id2
 ```go
 count, err := client.ExecuteRaw(`UPDATE "User" SET name = $1 WHERE id = $2`, "John", "123").Exec(ctx)
 ```
+
+## Next steps
+
+Check out the go client's [current limitations](./10-limitations.md).
```
Comment (matthewmueller) [inline docs/reference/09-raw.md:51]: go => Go

File: docs/reference/10-limitations.md [hunk 22]
```diff
@@ -0,0 +1,3 @@
+# Limitations
+
+Note that the go client is a prototype at this time and not ready for production use. It's missing a lot of features compared to the javascript client. However, in most of the cases, you can always [fall back to using raw queries](./09-raw.md).
```
Comment (matthewmueller) [inline docs/reference/10-limitations.md:3]: go => Go

Comment (matthewmueller) [reply docs/reference/10-limitations.md:3]: Should we say what is available is well-tested?

**Review by matthewmueller — CHANGES_REQUESTED** [review]
This is great. Nice work on this.

File: docs/reference/01-introduction.md (hunk 1 above)
Comment (matthewmueller) [reply docs/reference/01-introduction.md:5]: typesafe => type-safe

File: docs/reference/02-find.md (hunk 7 above)
Comment (steebchen) [reply docs/reference/02-find.md:39]: maybe ``err` returns an empty array`

File: docs/reference/02-find.md (hunk 8 above)
Comment (steebchen) [reply docs/reference/02-find.md:53]: I will do

> This returns an `ErrNotFound` error (exported by the generated client) if there was no such record.

File: docs/reference/02-find.md (hunk 9 above)
Comment (steebchen) [reply docs/reference/02-find.md:66]: image

File: docs/reference/03-filters.md (hunk 11 above)
Comment (steebchen) [reply docs/reference/03-filters.md:5]: > You probably want to build detailed queries, such as if a database column contains a word, [...]

?

File: docs/reference/01-introduction.md (hunk 5 above)
Comment (steebchen) [reply docs/reference/01-introduction.md:55]: to all parts of the chain, i.e. the model, the field and the action

Comment (steebchen) [reply docs/reference/01-introduction.md:55]: not sure what I would call that. make it's also obvious and we can just remove that part of the sentence?

File: docs/reference/03-filters.md (hunk 11 above)
Comment (matthewmueller) [reply docs/reference/03-filters.md]: On this page, we explore query filters. Query filters help us ...

... All of these queries are fully type-safe and independent of the underlying database.

**Review by matthewmueller — APPROVED** [review]

File: docs/reference/01-introduction.md (hunk 5 above)
Comment (steebchen) [reply docs/reference/01-introduction.md:55]: ok when you don't understand on the first read it's pretty much useless and just a bad idea from my side. will reword or remove
